dns-discovery/src/main/java/org/elasticsearch/discovery/srv/SrvUnicastHostsProvider.java

```diff
@@ -0,0 +1,161 @@
+
+    private final TransportService transportService;
+    private final Version version;
+    private final String query;
+    protected final Resolver resolver;
+
+    @Inject
+    public SrvUnicastHostsProvider(Settings settings, TransportService transportService, Version version) {
+        super(settings);
+        this.transportService = transportService;
+        this.version = version;
+        this.query = settings.get(DISCOVERY_SRV_QUERY);
+        this.resolver = resolver(settings);
+    }
+
+    @Nullable
+    private Resolver resolver(Settings settings) {
+        String hostname = null;
+        int port = -1;
+        String resolverAddress = settings.get(DISCOVERY_SRV_RESOLVER);
+        if (resolverAddress != null) {
+            String[] parts = resolverAddress.split(":");
+            if (parts.length > 0) {
+                hostname = parts[0];
+                if (parts.length > 1) {
+                    try {
+                        port = Integer.valueOf(parts[1]);
+                    } catch (Exception e) {
+                        logger.warn("Resolver port is '{}' not an integer", parts[1]);
+                    }
+                }
+
+            }
+        }
+        try {
+            Resolver resolver = new SimpleResolver(hostname);
+            if (port > 0) {
+                resolver.setPort(port);
+            }
+            return resolver;
+        } catch (UnknownHostException e) {
+            logger.warn("Could not create resolver. Using default resolver.", e);
+        }
+        return null;
+    }
+
+    @Override
+    public List<DiscoveryNode> buildDynamicNodes() {
+        List<DiscoveryNode> discoNodes = Lists.newArrayList();
+        if (query == null) {
+            logger.error("DNS query must not be null. Please set '{}'", DISCOVERY_SRV_QUERY);
+            return discoNodes;
+        }
+        try {
+            Record[] records = lookupRecords();
+            logger.trace("Building dynamic unicast discovery nodes...");
+            if (records == null || records.length == 0) {
+                logger.debug("No nodes found");
+            } else {
+                discoNodes = parseRecords(records);
+            }
+        } catch (TextParseException e) {
+            logger.error("Unable to parse DNS query '{}'", query);
+            logger.error("DNS lookup exception:", e);
+        }
+        logger.info("Using dynamic discovery nodes {}", discoNodes);
+        return discoNodes;
+    }
+
+    protected Record[] lookupRecords() throws TextParseException {
+        Lookup lookup = new Lookup(query, Type.SRV);
+        if (this.resolver != null) {
+            lookup.setResolver(this.resolver);
+        }
+        return lookup.run();
+    }
+
+    protected List<DiscoveryNode> parseRecords(Record[] records) {
+        ArrayList<Record> sortedList = Lists.newArrayList(records);
+        sortedList.sort(new Comparator<Record>() {
```

Comment:
buildDynamicNodes is called in ``UnicastZenPing`` and added to a set + sorted afterwards. So the sorting here won't have any effect and can be removed (unless it makes testing easier?)